File: ts/src/analytics/exploration-collapse-guard.ts
export type GuidanceKind =
  | "hint"
  | "playbook_update"
  | "teacher_signal"
  | "pressure_mode"
  | "other";
export type CollapseMetric =
  | "response_length"
  | "diversity"
  | "entropy"
  | "route_repetition"
  | "rollback_rate"
  | "score";
export type MitigationAction = "none" | "demote_guidance";

export type ExplorationSnapshot = {
  generationIndex: number;
  responseLength: number;
  diversity?: number;
  entropy?: number;
  routeSignature?: string;
  rollbackRate?: number;
  score?: number;
};

export type GuidanceChange = {
  changeId: string;
  generationIndex: number;
  kind: GuidanceKind;
  sourceComponent: string;
  sourceSpan?: string;
};

export type ExplorationCollapseThresholds = {
  window: number;
  minSignals: number;
  responseLengthDropRatio: number;
  diversityDropRatio: number;
  entropyDropRatio: number;
  routeRepetitionIncrease: number;
  rollbackRateIncrease: number;
  scoreDrop: number;
};

export type ExplorationCollapseSignal = {
  metric: CollapseMetric;
  before: number;
  after: number;
  delta: number;
  threshold: number;
};

export type ExplorationCollapseEvent = {
  eventType: "exploration_collapse_detected";
  guidanceChange: GuidanceChange;
  advisoryOnly: boolean;
  mitigation: MitigationAction;
  signals: ExplorationCollapseSignal[];
  recommendation: string;
};

export type ExplorationCollapseRecord = {
  event_type: "exploration_collapse_detected";
  payload: {
    guidance_change: {
      change_id: string;
      generation_index: number;
      kind: GuidanceKind;
      source_component: string;
      source_span?: string;
    };
    advisory_only: boolean;
    mitigation: MitigationAction;
    signals: Array<{
      metric: CollapseMetric;
      before: number;
      after: number;
      delta: number;
      threshold: number;
    }>;
    recommendation: string;
  };
};

export type ExplorationCollapseReport = {
  schemaVersion: 1;
  advisoryOnly: boolean;
  events: ExplorationCollapseEvent[];
  records: ExplorationCollapseRecord[];
};

export type ExplorationCollapseOptions = {
  advisoryOnly?: boolean;
  autoMitigation?: boolean;
  thresholds?: Partial<ExplorationCollapseThresholds>;
};

const DEFAULT_THRESHOLDS: ExplorationCollapseThresholds = {
  window: 2,
  minSignals: 2,
  responseLengthDropRatio: 0.25,
  diversityDropRatio: 0.25,
  entropyDropRatio: 0.25,
  routeRepetitionIncrease: 0.3,
  rollbackRateIncrease: 0.2,
  scoreDrop: 0.05,
};

export function detectExplorationCollapse(
  snapshots: ExplorationSnapshot[],
  guidanceChanges: GuidanceChange[],
  options: ExplorationCollapseOptions = {},
): ExplorationCollapseReport {
  const advisoryOnly = options.advisoryOnly ?? true;
  const thresholds = { ...DEFAULT_THRESHOLDS, ...options.thresholds };
  const ordered = [...snapshots].sort(
    (left, right) => left.generationIndex - right.generationIndex,
  );
  const events: ExplorationCollapseEvent[] = [];

  for (const change of guidanceChanges) {
    const before = ordered
      .filter((snapshot) => snapshot.generationIndex < change.generationIndex)
      .slice(-thresholds.window);
    const after = ordered
      .filter((snapshot) => snapshot.generationIndex >= change.generationIndex)
      .slice(0, thresholds.window);
    if (before.length === 0 || after.length === 0) continue;

    const signals = collapseSignals(before, after, thresholds);
    if (signals.length >= thresholds.minSignals) {
      events.push({
        eventType: "exploration_collapse_detected",
        guidanceChange: change,
        advisoryOnly,
        mitigation: options.autoMitigation && !advisoryOnly ? "demote_guidance" : "none",
        signals,
        recommendation: recommendation(Boolean(options.autoMitigation), advisoryOnly),
      });
    }
  }

  return {
    schemaVersion: 1,
    advisoryOnly,
    events,
    records: events.map(eventRecord),
  };
}

export function renderExplorationCollapseReport(report: ExplorationCollapseReport): string {
  if (report.events.length === 0) return "No exploration collapse detected.";
  const lines = ["# Exploration Collapse Guard", ""];
  for (const event of report.events) {
    const change = event.guidanceChange;
    const span = change.sourceSpan ? ` span=${change.sourceSpan}` : "";
    lines.push(
      `- ${change.changeId} (${change.kind}) at generation ${change.generationIndex}: ` +
        `source=${change.sourceComponent}${span}; ` +
        `metrics=${event.signals.map((signal) => signal.metric).join(", ")}; ` +
        `mitigation=${event.mitigation}.`,
    );
  }
  return `${lines.join("\n")}\n`;
}

function collapseSignals(
  before: ExplorationSnapshot[],
  after: ExplorationSnapshot[],
  thresholds: ExplorationCollapseThresholds,
): ExplorationCollapseSignal[] {
  const signals: ExplorationCollapseSignal[] = [];
  dropSignal(
    signals,
    "response_length",
    avg(before, (snapshot) => snapshot.responseLength),
    avg(after, (snapshot) => snapshot.responseLength),
    thresholds.responseLengthDropRatio,
  );
  dropSignal(
    signals,
    "diversity",
    avg(before, (snapshot) => snapshot.diversity),
    avg(after, (snapshot) => snapshot.diversity),
    thresholds.diversityDropRatio,
  );
  dropSignal(
    signals,
    "entropy",
    avg(before, (snapshot) => snapshot.entropy),
    avg(after, (snapshot) => snapshot.entropy),
    thresholds.entropyDropRatio,
  );
  riseSignal(
    signals,
    "route_repetition",
    routeRepetition(before),
    routeRepetition(after),
    thresholds.routeRepetitionIncrease,
  );
  riseSignal(
    signals,
    "rollback_rate",
    avg(before, (snapshot) => snapshot.rollbackRate),
    avg(after, (snapshot) => snapshot.rollbackRate),
    thresholds.rollbackRateIncrease,
  );
  absoluteDropSignal(
    signals,
    "score",
    avg(before, (snapshot) => snapshot.score),
    avg(after, (snapshot) => snapshot.score),
    thresholds.scoreDrop,
  );
  return signals;
}

function dropSignal(
  signals: ExplorationCollapseSignal[],
  metric: CollapseMetric,
  before: number | undefined,
  after: number | undefined,
  threshold: number,
): void {
  if (before === undefined || after === undefined || before <= 0) return;
  if ((before - after) / before >= threshold)
    appendSignal(signals, metric, before, after, threshold);
}

function absoluteDropSignal(
  signals: ExplorationCollapseSignal[],
  metric: CollapseMetric,
  before: number | undefined,
  after: number | undefined,
  threshold: number,
): void {
  if (before !== undefined && after !== undefined && before - after >= threshold) {
    appendSignal(signals, metric, before, after, threshold);
  }
}

function riseSignal(
  signals: ExplorationCollapseSignal[],
  metric: CollapseMetric,
  before: number | undefined,
  after: number | undefined,
  threshold: number,
): void {
  if (before !== undefined && after !== undefined && after - before >= threshold) {
    appendSignal(signals, metric, before, after, threshold);
  }
}

function appendSignal(
  signals: ExplorationCollapseSignal[],
  metric: CollapseMetric,
  before: number,
  after: number,
  threshold: number,
): void {
  signals.push({
    metric,
    before: round(before),
    after: round(after),
    delta: round(after - before),
    threshold,
  });
}

function avg(
  items: ExplorationSnapshot[],
  selector: (snapshot: ExplorationSnapshot) => number | undefined,
): number | undefined {
  const values = items
    .map(selector)
    .filter((value): value is number => typeof value === "number" && Number.isFinite(value));
  if (values.length === 0) return undefined;
  return round(values.reduce((total, value) => total + value, 0) / values.length);
}

function routeRepetition(items: ExplorationSnapshot[]): number | undefined {
  const routes = items
    .map((item) => item.routeSignature)
    .filter((route): route is string => Boolean(route));
  if (routes.length === 0) return undefined;
  const counts = new Map<string, number>();
  for (const route of routes) counts.set(route, (counts.get(route) ?? 0) + 1);
  return round(Math.max(...counts.values()) / routes.length);
}

function eventRecord(event: ExplorationCollapseEvent): ExplorationCollapseRecord {
  const change = event.guidanceChange;
  return {
    event_type: "exploration_collapse_detected",
    payload: {
      guidance_change: {
        change_id: change.changeId,
        generation_index: change.generationIndex,
        kind: change.kind,
        source_component: change.sourceComponent,
        ...(change.sourceSpan ? { source_span: change.sourceSpan } : {}),
      },
      advisory_only: event.advisoryOnly,
      mitigation: event.mitigation,
      signals: event.signals.map((signal) => ({ ...signal })),
      recommendation: event.recommendation,
    },
  };
}

function recommendation(autoMitigation: boolean, advisoryOnly: boolean): string {
  if (autoMitigation && !advisoryOnly) {
    return "Demote the associated guidance and switch to exploration-heavy sampling.";
  }
  return "Warn only; inspect the associated guidance before changing run behavior.";
}

function round(value: number): number {
  return Math.round(value * 1_000_000) / 1_000_000;
}

File: autocontext/src/autocontext/analytics/exploration_collapse_guard.py
"""Advisory guard for guidance-induced exploration collapse."""

from __future__ import annotations

import json
from collections import Counter
from math import isfinite
from pathlib import Path
from statistics import mean
from typing import Any, Literal, Self

from pydantic import BaseModel, ConfigDict, Field  # type: ignore[import-not-found]

GuidanceKind = Literal["hint", "playbook_update", "teacher_signal", "pressure_mode", "other"]
CollapseMetric = Literal[
    "response_length",
    "diversity",
    "entropy",
    "route_repetition",
    "rollback_rate",
    "score",
]
MitigationAction = Literal["none", "demote_guidance"]


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    def to_dict(self) -> dict[str, Any]:
        return self.model_dump(mode="json")

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Self:
        return cls.model_validate(data)


class ExplorationSnapshot(_StrictModel):
    generation_index: int = Field(ge=0)
    response_length: float = Field(ge=0)
    diversity: float | None = Field(default=None, ge=0.0)
    entropy: float | None = Field(default=None, ge=0.0)
    route_signature: str | None = None
    rollback_rate: float | None = Field(default=None, ge=0.0)
    score: float | None = None


class GuidanceChange(_StrictModel):
    change_id: str = Field(min_length=1)
    generation_index: int = Field(ge=0)
    kind: GuidanceKind
    source_component: str = Field(min_length=1)
    source_span: str | None = None


class ExplorationCollapseThresholds(_StrictModel):
    window: int = Field(default=2, ge=1)
    min_signals: int = Field(default=2, ge=1)
    response_length_drop_ratio: float = Field(default=0.25, ge=0.0, le=1.0)
    diversity_drop_ratio: float = Field(default=0.25, ge=0.0, le=1.0)
    entropy_drop_ratio: float = Field(default=0.25, ge=0.0, le=1.0)
    route_repetition_increase: float = Field(default=0.3, ge=0.0, le=1.0)
    rollback_rate_increase: float = Field(default=0.2, ge=0.0, le=1.0)
    score_drop: float = Field(default=0.05, ge=0.0)


class ExplorationCollapseSignal(_StrictModel):
    metric: CollapseMetric
    before: float
    after: float
    delta: float
    threshold: float


class ExplorationCollapseEvent(_StrictModel):
    event_type: Literal["exploration_collapse_detected"] = "exploration_collapse_detected"
    guidance_change: GuidanceChange
    advisory_only: bool
    mitigation: MitigationAction
    signals: list[ExplorationCollapseSignal]
    recommendation: str

    def to_record(self) -> dict[str, Any]:
        return {
            "event_type": self.event_type,
            "payload": {
                "guidance_change": self.guidance_change.to_dict(),
                "advisory_only": self.advisory_only,
                "mitigation": self.mitigation,
                "signals": [signal.to_dict() for signal in self.signals],
                "recommendation": self.recommendation,
            },
        }


class ExplorationCollapseReport(_StrictModel):
    schema_version: Literal[1] = 1
    advisory_only: bool
    events: list[ExplorationCollapseEvent]
    records: list[dict[str, Any]]


def detect_exploration_collapse(
    snapshots: list[ExplorationSnapshot],
    guidance_changes: list[GuidanceChange],
    *,
    advisory_only: bool = True,
    auto_mitigation: bool = False,
    thresholds: ExplorationCollapseThresholds | None = None,
) -> ExplorationCollapseReport:
    policy = thresholds or ExplorationCollapseThresholds()
    ordered = sorted(snapshots, key=lambda item: item.generation_index)
    events: list[ExplorationCollapseEvent] = []
    for change in guidance_changes:
        before = [item for item in ordered if item.generation_index < change.generation_index][-policy.window :]
        after = [item for item in ordered if item.generation_index >= change.generation_index][: policy.window]
        if not before or not after:
            continue
        signals = _signals(before, after, policy)
        if len(signals) >= policy.min_signals:
            events.append(
                ExplorationCollapseEvent(
                    guidance_change=change,
                    advisory_only=advisory_only,
                    mitigation="demote_guidance" if auto_mitigation and not advisory_only else "none",
                    signals=signals,
                    recommendation=_recommendation(auto_mitigation, advisory_only),
                )
            )
    return ExplorationCollapseReport(
        advisory_only=advisory_only,
        events=events,
        records=[event.to_record() for event in events],
    )


def persist_exploration_collapse_report(path: Path, report: ExplorationCollapseReport) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "schema_version": report.schema_version,
        "advisory_only": report.advisory_only,
        "events": report.records,
    }
    path.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def render_exploration_collapse_report(report: ExplorationCollapseReport) -> str:
    if not report.events:
        return "No exploration collapse detected."
    lines = ["# Exploration Collapse Guard", ""]
    for event in report.events:
        change = event.guidance_change
        metrics = ", ".join(signal.metric for signal in event.signals)
        span = f" span={change.source_span}" if change.source_span else ""
        lines.append(
            f"- {change.change_id} ({change.kind}) at generation {change.generation_index}: "
            f"source={change.source_component}{span}; metrics={metrics}; mitigation={event.mitigation}."
        )
    return "\n".join(lines) + "\n"


def _signals(
    before: list[ExplorationSnapshot],
    after: list[ExplorationSnapshot],
    thresholds: ExplorationCollapseThresholds,
) -> list[ExplorationCollapseSignal]:
    signals: list[ExplorationCollapseSignal] = []
    _drop_signal(
        signals,
        "response_length",
        _avg(before, "response_length"),
        _avg(after, "response_length"),
        thresholds.response_length_drop_ratio,
    )
    _drop_signal(signals, "diversity", _avg(before, "diversity"), _avg(after, "diversity"), thresholds.diversity_drop_ratio)
    _drop_signal(signals, "entropy", _avg(before, "entropy"), _avg(after, "entropy"), thresholds.entropy_drop_ratio)
    _rise_signal(
        signals,
        "route_repetition",
        _route_repetition(before),
        _route_repetition(after),
        thresholds.route_repetition_increase,
    )
    _rise_signal(
        signals,
        "rollback_rate",
        _avg(before, "rollback_rate"),
        _avg(after, "rollback_rate"),
        thresholds.rollback_rate_increase,
    )
    _absolute_drop_signal(signals, "score", _avg(before, "score"), _avg(after, "score"), thresholds.score_drop)
    return signals


def _drop_signal(
    signals: list[ExplorationCollapseSignal],
    metric: CollapseMetric,
    before: float | None,
    after: float | None,
    threshold: float,
) -> None:
    if before is None or after is None or before <= 0:
        return
    ratio = (before - after) / before
    if ratio >= threshold:
        _append_signal(signals, metric, before, after, threshold)


def _absolute_drop_signal(
    signals: list[ExplorationCollapseSignal],
    metric: CollapseMetric,
    before: float | None,
    after: float | None,
    threshold: float,
) -> None:
    if before is not None and after is not None and before - after >= threshold:
        _append_signal(signals, metric, before, after, threshold)


def _rise_signal(
    signals: list[ExplorationCollapseSignal],
    metric: CollapseMetric,
    before: float | None,
    after: float | None,
    threshold: float,
) -> None:
    if before is not None and after is not None and after - before >= threshold:
        _append_signal(signals, metric, before, after, threshold)


def _append_signal(
    signals: list[ExplorationCollapseSignal],
    metric: CollapseMetric,
    before: float,
    after: float,
    threshold: float,
) -> None:
    signals.append(
        ExplorationCollapseSignal(
            metric=metric,
            before=before,
            after=after,
            delta=after - before,
            threshold=threshold,
        )
    )


def _avg(items: list[ExplorationSnapshot], attr: str) -> float | None:
    values = [float(value) for item in items if (value := getattr(item, attr)) is not None and isfinite(float(value))]
    return round(mean(values), 6) if values else None


def _route_repetition(items: list[ExplorationSnapshot]) -> float | None:
    routes = [item.route_signature for item in items if item.route_signature]
    if not routes:
        return None
    return round(Counter(routes).most_common(1)[0][1] / len(routes), 6)


def _recommendation(auto_mitigation: bool, advisory_only: bool) -> str:
    if auto_mitigation and not advisory_only:
        return "Demote the associated guidance and switch to exploration-heavy sampling."
    return "Warn only; inspect the associated guidance before changing run behavior."
